src/bijectors/stacked.jl

```diff
@@ -230,15 +230,16 @@ end
 end
 
 function _with_logabsdet_jacobian(sb::Stacked, x::AbstractVector)
-    N = length(sb.bs)
-    yinit, linit = with_logabsdet_jacobian(sb.bs[1], x[sb.ranges_in[1]])
-    logjac = sum(linit)
-    ys = mapreduce(vcat, sb.bs[2:end], sb.ranges_in[2:end]; init=yinit) do b, r
-        y, l = with_logabsdet_jacobian(b, x[r])
-        logjac += sum(l)
-        y
-    end
-    return (ys, logjac)
+    ys = map(zip(sb.bs, sb.ranges_in)) do tup
```

Comment:
Isn't separating these two computations a bit of an overkill? As in, why not do

```julia
ys_and_logjacs = map(zip(sb.bs sb.ranges_in)) do (b, r)
    y, logjac = with_logabsdet_jacobian(b, x[r])
    (y, logjac)
end
y = reduce(vcat, map(first, ys_and_logjacs))
logjac = sum(map(last, ys_and_logjacs))
```

?

Note also that I _believe_ certain AD backends, e.g. Zygote.jl, perform much better if you do something like `sum(xs)` rather than `sum(f, xs)`. Don't have a reference now, but probably is just the specialization thingy discussed elsewhere.

Reply:
Indeed yeah :sweat: . But the less-overkill solution seems to break Zygote. Have no idea why.

```julia
julia> main()
ERROR: MethodError: no method matching +(::Base.RefValue{Any}, ::@NamedTuple{contents::Vector{Float64}})

Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...)
   @ Base operators.jl:587
  +(::ChainRulesCore.ZeroTangent, ::Any)
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/I1EbV/src/tangent_arithmetic.jl:99
  +(::Any, ::ChainRulesCore.NotImplemented)
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/I1EbV/src/tangent_arithmetic.jl:25
  ...

Stacktrace:
  [1] accum(x::Base.RefValue{Any}, y::@NamedTuple{contents::Vector{Float64}})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/lib/lib.jl:17
  [2] accum(x::Base.RefValue{Any}, y::Nothing, zs::@NamedTuple{contents::Vector{Float64}})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/lib/lib.jl:22
  [3] _with_logabsdet_jacobian
    @ ~/Projects/advancedvi.jl:77 [inlined]
  [4] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Vector{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
  [5] with_logabsdet_jacobian
    @ ~/.julia/dev/Bijectors/src/bijectors/stacked.jl:249 [inlined]
  [6] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Vector{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
  [7] #1
    @ ~/.julia/dev/AdvancedVI/ext/AdvancedVIBijectorsExt.jl:39 [inlined]
  [8] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Vector{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
  [9] MappingRF
    @ ./reduce.jl:100 [inlined]
 [10] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [11] _foldl_impl
    @ ./reduce.jl:62 [inlined]
 [12] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [13] foldl_impl
    @ ./reduce.jl:48 [inlined]
 [14] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [15] mapfoldl_impl
    @ ./reduce.jl:44 [inlined]
 [16] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [17] #mapfoldl#298
    @ ./reduce.jl:175 [inlined]
 [18] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [19] mapfoldl
    @ ./reduce.jl:175 [inlined]
 [20] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [21] #mapreduce#302
    @ ./reduce.jl:307 [inlined]
 [22] (::Zygote.Pullback{Tuple{…}, Any})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [23] mapreduce
    @ ./reduce.jl:307 [inlined]
 [24] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [25] reparam_with_entropy
    @ ~/.julia/dev/AdvancedVI/ext/AdvancedVIBijectorsExt.jl:34 [inlined]
 [26] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Tuple{Matrix{…}, Float64})
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [27] f
    @ ~/.julia/dev/AdvancedVI/src/objectives/elbo/repgradelbo.jl:110 [inlined]
 [28] (::Zygote.Pullback{Tuple{…}, Tuple{…}})(Δ::Float64)
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
 [29] (::Zygote.var"#75#76"{Zygote.Pullback{Tuple{…}, Tuple{…}}})(Δ::Float64)
    @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface.jl:91
 [30] value_and_gradient!(ad::AutoZygote, f::Function, θ::Vector{…}, out::DiffResults.MutableDiffResult{…})
    @ AdvancedVIZygoteExt ~/.julia/dev/AdvancedVI/ext/AdvancedVIZygoteExt.jl:18
 [31] estimate_gradient!(rng::Random.TaskLocalRNG, obj::RepGradELBO{…}, adtype::AutoZygote, out::DiffResults.MutableDiffResult{…}, prob::Target{…}, λ::Vector{…}, restructure::Optimisers.Restructure{…}, state::Nothing)
    @ AdvancedVI ~/.julia/dev/AdvancedVI/src/objectives/elbo/repgradelbo.jl:115
```

Reply:
Ah 

```julia
function _with_logabsdet_jacobian(sb::Stacked, x::AbstractVector)
    ys_and_logjacs = map(zip(sb.bs, sb.ranges_in)) do (b, r)
        with_logabsdet_jacobian(b, x[r])
    end
    ys      = map(first, ys_and_logjacs)
    logjacs = map(last, ys_and_logjacs)
    y       = reduce(vcat, ys)
    logjac  = sum(logjacs)
    y, logjac
end
```
This works. It seems like the compiler insists on unpacking and then repacking the Tuple, which seems to cause some type instability nonsense? Quite bizarre that the compiler doesn't skip the packing/unpacking.

Reply:
Neato! But you need to use proper formatting (you can see the JuliaFormatter is failing).

Also, generally speaking, I'd suggest following standard formatting, i.e. according to JuliaFormatter everywhere, e.g. AdvancedVI.jl too; keeps things consistent:)

Reply:
I was being really confused because I thought I followed the formatting and the formatter was indeed doing nothing: trailing whitespace :sweat: